cmd/nerdctl/run_test.go

```diff
@@ -449,3 +449,76 @@ COPY --from=builder /go/src/logger/logger /
 	assert.Check(t, strings.Contains(log, "foo"))
 	assert.Check(t, strings.Contains(log, "bar"))
 }
+
+// this container only works properly with -t -d (or -t) modes
+const TerminalContainerDockerfile = `
+FROM golang:latest as builder
+WORKDIR /go/src/
+RUN mkdir -p app
+WORKDIR /go/src/app
+RUN echo '\
+	package main \n\
+	import ( \n\
+		"fmt" \n\
+		"time" \n\
+	\n\
+		"github.com/containerd/console" \n\
+	) \n\
+	\n\
+	func main() { \n\
+		con := console.Current() \n\
+		defer con.Reset() \n\
+		fmt.Println(con) \n\
+		for { \n\
+			con.Write([]byte(time.Now().String() + "\\n")) \n\
+			time.Sleep(time.Second) \n\
+		} \n\
+	} \n' >> app.go
+
+RUN go mod init
+RUN go mod tidy
+RUN go build .
+
+FROM alpine:3.13
+COPY --from=builder /go/src/app/app /root
+RUN chmod +x /root/app
+CMD ["/root/app"]
+	`
+
+func TestNewTerminalDetachedMode(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("buildkit is not enabled on windows, this feature may work on windows.")
+	}
+	testutil.DockerIncompatible(t)
+	t.Parallel()
+	base := testutil.NewBase(t)
+	imageName := testutil.Identifier(t) + "-image"
+	containerName := testutil.Identifier(t)
+
+	buildCtx, err := createBuildContext(TerminalContainerDockerfile)
+	assert.NilError(t, err)
+	defer os.RemoveAll(buildCtx)
+
+	base.Cmd("build", "-t", imageName, buildCtx).AssertOK()
+	defer base.Cmd("image", "rm", "-f", imageName).AssertOK()
+
+	// without -t, fail
+	base.Cmd("run", "-d", "--name", containerName, imageName).AssertOK()
+	base.Cmd("ps", "-a").AssertOutWithFunc(func(stdout string) error {
+		if !strings.Contains(stdout, "Exited (2)") {
+			return fmt.Errorf("terminal container run success without -t")
+		}
+		return nil
+	})
+	base.Cmd("container", "rm", "-f", containerName).AssertOK()
+
+	// with -t, success
+	base.Cmd("run", "-d", "-t", "--name", containerName, imageName).AssertOK()
+	base.Cmd("ps", "-a").AssertOutWithFunc(func(stdout string) error {
+		if !strings.Contains(stdout, "Up") {
+			return fmt.Errorf("terminal container run fail with -t")
+		}
+		return nil
+	})
+	base.Cmd("container", "rm", "-f", containerName).AssertOK()
```

Comment:
This line can be a `defer` and moved up to before or after line 506, so that the `rm -f containerName` is always called.

Right now if either 507 or 517 fails, the container won't be cleaned.

Reply:
Here are actually two little tests, with "-t" and without "-t", so the container must be deleted before running again. 
Perhaps use different name, or test separately?

Reply:
I think you can use different names and `defer` both `rm -f containerName` calls.

Same for the test in [cmd/nerdctl/create_linux_test.go](https://github.com/containerd/nerdctl/pull/1642/files#diff-fcc7da15472e79a3ccbd7e3c71d107994cc6d7b5f3efec8b5fe50e38ba36eae1)

Reply:
> I think you can use different names and `defer` both `rm -f containerName` calls.
> 
> Same for the test in [cmd/nerdctl/create_linux_test.go](https://github.com/containerd/nerdctl/pull/1642/files#diff-fcc7da15472e79a3ccbd7e3c71d107994cc6d7b5f3efec8b5fe50e38ba36eae1)

Great, thanks